stix2/sources/taxii.py

```diff
@@ -10,66 +10,125 @@
 
 """
 
-import json
-
+from stix2.base import _STIXBase
+from stix2.core import Bundle, parse
 from stix2.sources import DataSink, DataSource, DataStore, make_id
 from stix2.sources.filters import Filter
 
 TAXII_FILTERS = ['added_after', 'id', 'type', 'version']
 
 
 class TAXIICollectionStore(DataStore):
-    """
-    """
-    def __init__(self, collection):
-        """
-        Create a new TAXII Collection Data store
+    """TAXIICollectionStore
 
-        Args:
-            collection (taxii2.Collection): Collection instance
+    Provides an interface to a local/remote TAXII Collection
+    of STIX data. TAXIICollectionStore is a wrapper
+    around a paired TAXIICollectionSink and TAXIICollectionSource.
 
-        """
+    Args:
+            collection (taxii2.Collection): TAXII Collection instance
+    """
+    def __init__(self, collection):
         super(TAXIICollectionStore, self).__init__()
         self.source = TAXIICollectionSource(collection)
         self.sink = TAXIICollectionSink(collection)
 
 
 class TAXIICollectionSink(DataSink):
-    """
+    """TAXIICollectionSink
+
+    Provides an interface for pushing STIX objects to a local/remote
+    TAXII Collection endpoint.
+
+    Args:
+        collection (taxii2.Collection): TAXII2 Collection instance
+
     """
     def __init__(self, collection):
         super(TAXIICollectionSink, self).__init__()
         self.collection = collection
 
-    def add(self, stix_obj):
-        """
+    def add(self, stix_data):
+        """add/push STIX content to TAXII Collection endpoint
+
+        Args:
+            stix_data (STIX object OR dict OR str OR list): valid STIX 2.0 content
+                in a STIX object (or Bundle), STIX onject dict (or Bundle dict), or a STIX 2.0
+                json encoded string, or list of any of the following
+
         """
-        self.collection.add_objects(self.create_bundle([json.loads(str(stix_obj))]))
+
+        if isinstance(stix_data, _STIXBase):
+            # adding python STIX object
+            bundle = dict(Bundle(stix_data))
+
+        elif isinstance(stix_data, dict):
+            # adding python dict (of either Bundle or STIX obj)
+            if stix_data["type"] == "bundle":
+                bundle = stix_data
+            else:
+                bundle = dict(Bundle(stix_data))
+
+        elif isinstance(stix_data, list):
+            # adding list of something - recurse on each
+            for obj in stix_data:
+                self.add(obj)
+        elif isinstance(stix_data, str):
+            # adding json encoded string of STIX content
+            stix_data = parse(stix_data)
+            if stix_data["type"] == "bundle":
+                bundle = dict(stix_data)
+            else:
+                bundle = dict(Bundle(stix_data))
+
+        self.collection.add_objects(bundle)
 
     @staticmethod
     def create_bundle(objects):
+        """TODO: Remove?"""
```

Comment:
Yes, this should be removed - it duplicates the `Bundle` constructor.

Reply:
removed